File: tan-core/src/main/java/org/leralix/tan/gui/cosmetic/type/NexoIconType.java
package org.leralix.tan.gui.cosmetic.type;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.leralix.tan.integration.NexoIntegration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Icon type that loads Nexo custom items at runtime.
 * Falls back to Material.BARRIER if Nexo unavailable or item not found.
 *
 * <p>Usage in icons.yml:
 * <pre>{@code
 * TERRITORY_ICON: "nexo:custom_crown"
 * TREASURY_ICON: "nexo:gold_coin_stack"
 * WAR_ICON: "nexo:enchanted_sword"
 * }</pre>
 */
public class NexoIconType extends IconType {

    private static final Logger logger = LoggerFactory.getLogger(NexoIconType.class);
    
    private final String nexoItemId;
    private final Material fallbackMaterial;
    private ItemStack cachedItem = null;

    /**
     * Create a Nexo icon type with BARRIER as fallback.
     *
     * @param nexoItemId the Nexo item ID (without "nexo:" prefix)
     */
    public NexoIconType(String nexoItemId) {
        this(nexoItemId, Material.BARRIER);
    }

    /**
     * Create a Nexo icon type with custom fallback material.
     *
     * @param nexoItemId the Nexo item ID (without "nexo:" prefix)
     * @param fallbackMaterial the Material to use if Nexo unavailable
     */
    public NexoIconType(String nexoItemId, Material fallbackMaterial) {
        this.nexoItemId = nexoItemId;
        this.fallbackMaterial = fallbackMaterial;
    }

    @Override
    protected ItemStack getItemStack(Player player) {
        // Cache the item to avoid repeated Nexo API calls
        if (cachedItem != null) {
            return cachedItem.clone();
        }

        // Try to load from Nexo
        ItemStack nexoItem = NexoIntegration.getItemOrDefault(
            nexoItemId, 
            new ItemStack(fallbackMaterial)
        );

        // Cache successful load
        if (nexoItem.getType() != fallbackMaterial) {
            cachedItem = nexoItem;
            logger.debug("Cached Nexo item '{}'", nexoItemId);
        }

        return nexoItem.clone();
    }

    /**
     * Get the Nexo item ID.
     *
     * @return the item ID without prefix
     */
    public String getNexoItemId() {
        return nexoItemId;
    }

    /**
     * Get the fallback material.
     *
     * @return the Material used when Nexo unavailable
     */
    public Material getFallbackMaterial() {
        return fallbackMaterial;
    }
}

File: tan-core/src/main/kotlin/org/leralix/tan/gui/cosmetic/type/NexoIconType.kt
package org.leralix.tan.gui.cosmetic.type
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.leralix.tan.integration.nexo.NexoIntegration
import org.slf4j.LoggerFactory
class NexoIconType(
    val nexoItemId: String,
    val fallbackMaterial: Material = Material.BARRIER
) : IconType() {
    companion object {
        private val logger = LoggerFactory.getLogger(NexoIconType::class.java)
        @JvmStatic
        fun parse(iconString: String): NexoIconType? {
            if (!iconString.startsWith("nexo:", ignoreCase = true)) {
                return null
            }
            val parts = iconString.split(":", limit = 3)
            if (parts.size < 2) {
                logger.warn("[NexoIconType] Invalid format (missing item ID): $iconString")
                return null
            }
            val itemId = parts[1].trim()
            if (itemId.isEmpty()) {
                logger.warn("[NexoIconType] Invalid format (empty item ID): $iconString")
                return null
            }
            val fallback = if (parts.size >= 3) {
                val fallbackName = parts[2].trim()
                try {
                    Material.valueOf(fallbackName)
                } catch (e: IllegalArgumentException) {
                    logger.warn("[NexoIconType] Invalid fallback material '$fallbackName' - using BARRIER")
                    Material.BARRIER
                }
            } else {
                Material.BARRIER
            }
            return NexoIconType(itemId, fallback)
        }
    }
    override fun getItemStack(player: Player?): ItemStack {
        val startTime = System.currentTimeMillis()
        val item = NexoIntegration.getItemOrDefault(nexoItemId, fallbackMaterial)
        val elapsed = System.currentTimeMillis() - startTime
        if (elapsed > 50) {
            logger.warn("[NexoIconType] Slow load for '{}' ({}ms) - consider pre-fetching", nexoItemId, elapsed)
        }
        val meta = item.itemMeta
        val hasCmd = meta?.hasCustomModelData() == true
        val cmd = if (hasCmd) meta?.customModelData else null
        logger.info("[NexoIconType] Loading '{}': type={}, hasCmd={}, cmd={}, isFallback={}",
            nexoItemId, item.type, hasCmd, cmd, item.type == fallbackMaterial)
        val cloned = item.clone()
        val clonedMeta = cloned.itemMeta
        val clonedHasCmd = clonedMeta?.hasCustomModelData() == true
        val clonedCmd = if (clonedHasCmd) clonedMeta?.customModelData else null
        logger.info("[NexoIconType] After clone '{}': hasCmd={}, cmd={}", nexoItemId, clonedHasCmd, clonedCmd)
        return cloned
    }
    fun isUsingFallback(): Boolean {
        if (!NexoIntegration.isEnabled) return true
        return !NexoIntegration.itemExists(nexoItemId)
    }
    override fun toString(): String {
        return "NexoIconType(itemId='$nexoItemId', fallback=$fallbackMaterial, usingFallback=${isUsingFallback()})"
    }
}